source/ecs/entity.d

```diff
@@ -9,214 +9,160 @@ import std.typecons : Nullable;
- * \
- * `Entity!ubyte` = **0000 0000** = ***(batch << 4) | id***.
- * \
- * What this means is that for a given value of `ubyte` it's first half is
- *     composed with the **id** and it's second half with the **batch**. This
- *     allows entities to be reused at some time in the program's life without
- *     having to resort to a more complicated process. Every time an entity is
- *     **discarded** it's **id** doesn't suffer any alterations however it's
- *     **batch** is increased by **1**, allowing the usage of an entity with the
- *     the same **id** but mantaining it's uniqueness with a new **batch**
- *     generating a completely new signature.
- *
- * See_Also: [skypjack - entt](https://skypjack.github.io/2019-05-06-ecs-baf-part-3/)
- */
-struct Entity(T)
-	if (isEntityType!T)
-{
-public:
-	this(in T id) { _id = id; }
-	this(in T id, in T batch) { _id = id; _batch = batch; }
 
 	@safe pure
-	bool opEquals(in Entity!T other) const
+	bool opEquals(in Entity other) const
 	{
 		return other.signature == signature;
 	}
 
+
 	@property @safe pure
-	T id() const { return _id; }
+	size_t id() const { return _id; }
+
 
 	@property @safe pure
-	T batch() const { return _batch; }
+	size_t batch() const { return _batch; }
+
 
 	@safe pure
 	auto incrementBatch()
 	{
-		_batch = _batch >= EntityManager!(T).batchMask ? 0 : cast(T)(_batch + 1);
+		_batch = _batch >= maxbatch ? 0 : _batch + 1;
 
 		return this;
 	}
 
 	@safe pure
-	T signature() const
+	size_t signature() const
+	{
+		return (_id | (_batch << idshift));
+	}
+
+	static if (typeof(int.sizeof).sizeof == 4)
 	{
-		return cast(T)(_id | (_batch << EntityManager!(T).entityShift));
+		enum size_t idshift = 20UL;   // 20 bits for ids
+		enum size_t maxid = 0xFFFF_F; // 1_048_575 unique ids
+		enum size_t maxbatch = 0xFFF; // 4_095 unique batches
 	}
+	else static if (typeof(int.sizeof).sizeof == 8)
+	{
+		enum size_t idshift = 32UL;      // 32 bits for ids
+		enum size_t maxid = 0xFFFF_FFFF; // 4_294_967_295 unique ids
+		enum size_t maxbatch = maxid;    // 4_294_967_295 unique batches
+	}
+	else
+		static assert(false, "unsuported target");
+
+	enum size_t idmask = maxid; // first 20 bits (sizeof==4), 32 bits (sizeof==8)
+	enum size_t batchmask = maxbatch << idshift; // last 12 bits (sizeof==4), 32 bits (sizeof==8)
+
+	alias signature this;
 
 private:
-	T _id;
-	T _batch;
+	size_t _id;
+	size_t _batch;
 }
 
 @safe pure
 @("entity: Entity")
 unittest
 {
-	auto entity0 = Entity!ubyte(0);
+	auto entity0 = Entity(0);
 
 	assertEquals(0, entity0.id);
 	assertEquals(0, entity0.batch);
 	assertEquals(0, entity0.signature);
-	assertEquals(Entity!ubyte(0, 0), entity0);
+	assertEquals(Entity(0, 0), entity0);
 
 	entity0.incrementBatch();
 	assertEquals(0, entity0.id);
 	assertEquals(1, entity0.batch);
-	assertEquals(16, entity0.signature);
-	assertEquals(Entity!ubyte(0, 1), entity0);
 
-	entity0 = Entity!ubyte(0, 15);
+	static if (typeof(int.sizeof).sizeof == 4)
```

Comment:
instead of calculating the sizeof a pointer, you can setup a private enum and use it internally. If you use this too many times on the code, code improve compilation times a little bit. Maybe not significative for the effort.

Reply:
This only affects the Entity, so there is no need for it.